code_map_support.py

```diff
@@ -0,0 +1,404 @@
+            "cols": [0.0, 1.0],
+            "rows": [0.0, 1.0]
+        })
+
+    elif count == 2:
+        sublime.active_window().run_command("set_layout", {
+            "cells": [[0, 0, 1, 1], [1, 0, 2, 1]],
+            "cols": [0.0, coll_width, 1.0],
+            "rows": [0.0, 1.0]
+        })
+
+    elif count == 3:
+        sublime.active_window().run_command("set_layout", {
+            "cells": [[0, 0, 1, 1], [1, 0, 2, 1], [2, 0, 3, 1]],
+            "cols": [0.0, 0.33, 0.66, 1.0],
+            "rows": [0.0, 1.0]
+        })
+
+    elif count == 4:
+        sublime.active_window().run_command("set_layout", {
+            "cells": [[0, 0, 1, 1], [1, 0, 2, 1], [2, 0, 3, 1], [3, 0, 4, 1]],
+            "cols": [0.0, 0.25, 0.5, 0.75, 1.0],
+            "rows": [0.0, 1.0]
+        })
+
+# -----------------
+
+
+def block_max_pane(mode):
+    """Prevent Max Pane from managing windows, unblock after PUI is closed"""
+
+    try:
+        s = sublime.load_settings("max_pane_share.sublime-settings")
+        s.set('block_max_pane', mode)
+    except:
+        return
+
+# ===============================================================================
+
+
+class NavigateCodeMap():
+
+    def highlight_line(v):
+        line = v.line(v.sel()[0].a)
+        v.sel().clear()
+        v.sel().add(sublime.Region(line.a, line.b))
+
+    def keep_going(v, up=False, down=False):
+        if up:
+            line = v.line(v.sel()[0].a-1)
+            s = "going up, "
+            if line.a == 0:
+                return False
+        else:
+            line = v.line(v.sel()[0].b+1)
+            s = "going down, "
+            if line.b == v.size():
+                return False
+        indent = v.substr(sublime.Region(line.a, line.a+1)) in [" ", r"\t"]
+        empty = line.empty()
+        return indent or empty
+
+    def up(v, fast):
+        v.run_command("move_to", {"to": "bol", "extend": False})
+        if not fast:
+            v.run_command("move", {"by": "lines", "forward": False})
+            if v.line(v.sel()[0]).empty():
+                v.run_command("move", {"by": "lines", "forward": False})
+        else:
+            x = NavigateCodeMap.keep_going(v, up=True)
+            while x is not False:
+                x = NavigateCodeMap.keep_going(v, up=True)
+                v.run_command("move", {"by": "lines", "forward": False})
+
+    def down(v, fast):
+        v.run_command("move_to", {"to": "bol", "extend": False})
+        v.run_command("move", {"by": "lines", "forward": True})
+        if not fast:
+            if v.line(v.sel()[0]).empty():
+                v.run_command("move", {"by": "lines", "forward": True})
+        else:
+            x = NavigateCodeMap.keep_going(v, down=True)
+            while x is not False:
+                x = NavigateCodeMap.keep_going(v, down=True)
+                v.run_command("move", {"by": "lines", "forward": True})
+            # v.run_command("move", {"by": "lines", "forward": True})
+
```

Comment:
this causes a crash I must fix, when moving fast (but not slow) over the end of file